feat: Add user management daemon

manoharan-nexthop opened the pull request:
This implementation addresses the User Management HLD requirements for centralized user administration in SONiC. https://github.com/sonic-net/SONiC/pull/2018

**1. User Management Daemon (userd):**
- Implemented Rust daemon using SWSS framework for CONFIG_DB integration
- Added user lifecycle management (create/update/delete/enable/disable)
- Implemented role-based group assignment (administrator, operator roles)
- Added SSH key management with proper file permissions and ownership
- Used posix_spawn() for secure command execution without shell interpretation
- Integrated PAM faillock configuration using Jinja2 templates

**2. Security Features:**
- Secure password handling using system's native hashing methods
- Role-based group assignments with predefined security groups
- PAM faillock integration for login attempt limiting

manoharan-nexthop committed:
[sonic-host-services] Add comprehensive user management daemon (userd)

SONiC needed a centralized user management daemon that can:
- Monitor CONFIG_DB for user configuration changes
- Manage local system users based on SONiC configuration
- Provide role-based access control with predefined groups
- Handle SSH key management with proper security
- Integrate with PAM for authentication policies
- Support efficient change detection to minimize system calls

**1. User Management Daemon (userd):**
- Implemented C++ daemon using SWSS framework for CONFIG_DB integration
- Added comprehensive user lifecycle management (create/update/delete/enable/disable)
- Implemented role-based group assignment (administrator, operator roles)
- Added SSH key management with proper file permissions and ownership
- Used posix_spawn() for secure command execution without shell interpretation
- Added efficient change detection using UserInfo comparison to avoid unnecessary system calls
- Integrated PAM faillock configuration using Jinja2 templates

**2. Build System Integration:**
- Added CMakeLists.txt for C++ compilation with SWSS dependencies
- Created debian packaging with proper control files and dependencies
- Added systemd service configuration for userd daemon
- Integrated Makefile for building and installation

**3. Security Features:**
- Secure password handling using system's native hashing methods
- Proper file permissions for SSH keys (600) and directories (700)
- Role-based group assignments with predefined security groups
- PAM faillock integration for login attempt limiting
- Input validation and sanitization for all user operations

**4. Testing Framework:**
- Added comprehensive unit tests for userd functionality
- Integration tests for CONFIG_DB interaction
- User lifecycle testing with proper cleanup
- SSH key management testing
- Role-based access control validation

Signed-off-by: Manoharan Sundaramoorthy <manoharan@nexthop.ai>

manoharan-nexthop committed:
[sonic-host-services] Add comprehensive user management daemon (userd)

SONiC needed a centralized user management daemon that can:
- Monitor CONFIG_DB for user configuration changes
- Manage local system users based on SONiC configuration
- Provide role-based access control with predefined groups
- Handle SSH key management with proper security
- Integrate with PAM for authentication policies
- Support efficient change detection to minimize system calls

**1. User Management Daemon (userd):**
- Implemented C++ daemon using SWSS framework for CONFIG_DB integration
- Added comprehensive user lifecycle management (create/update/delete/enable/disable)
- Implemented role-based group assignment (administrator, operator roles)
- Added SSH key management with proper file permissions and ownership
- Used posix_spawn() for secure command execution without shell interpretation
- Added efficient change detection using UserInfo comparison to avoid unnecessary system calls
- Integrated PAM faillock configuration using Jinja2 templates

**2. Build System Integration:**
- Added CMakeLists.txt for C++ compilation with SWSS dependencies
- Created debian packaging with proper control files and dependencies
- Added systemd service configuration for userd daemon
- Integrated Makefile for building and installation

**3. Security Features:**
- Secure password handling using system's native hashing methods
- Proper file permissions for SSH keys (600) and directories (700)
- Role-based group assignments with predefined security groups
- PAM faillock integration for login attempt limiting
- Input validation and sanitization for all user operations

**4. Testing Framework:**
- Added comprehensive unit tests for userd functionality
- Integration tests for CONFIG_DB interaction
- User lifecycle testing with proper cleanup
- SSH key management testing
- Role-based access control validation

Signed-off-by: Manoharan Sundaramoorthy <manoharan@nexthop.ai>

mssonicbld commented:
/azp run

manoharan-nexthop commented on data/debian/control:
```diff
@@ -3,7 +3,7 @@ Maintainer: Joe LeVeque <jolevequ@microsoft.com>
 Priority: optional
 Standards-Version: 0.1
-Build-Depends: debhelper (>=11)
+Build-Depends: debhelper (>=11), cmake, g++, pkg-config
```
No they are not needed, I seem to wrongly pushed some old changes. I have updated the right one now..

manoharan-nexthop commented on debian/control:
```diff
@@ -0,0 +1,11 @@
+Section: misc
+Priority: optional
+Standards-Version: 0.1
+Build-Depends: debhelper (>=11), cmake, g++, pkg-config, libcrypt-dev, nlohmann-json3-dev
```
same as above. I have updated the latest changes. It should have the right dependencies now.

rd-foredge commented on data/templates/faillock.conf.j2:
```diff
@@ -0,0 +1,27 @@
+# Role-based security policies
+{% for role, policy in security_policies.items() %}
+{% if policy.max_login_attempts %}
+# {{ role }} role settings
```
does faillock support individual config per group/role/user ?

manoharan-nexthop committed:
(userd) Replace C++ userd with Rust implementation

- Add userd-rs crate with full user management functionality
- Remove C++ userd implementation (userd/ directory)
- Add userd binary to sonic-host-services-rs package
- Add userd.service to sonic-host-services-data package

Features:
- User CRUD operations (create, update, delete)
- Role-based access control (administrator, operator)
- SSH key management with validation
- Security policies and PAM faillock configuration
- Consistency checking on startup
- CONFIG_DB monitoring via SubscriberStateTable

The Rust implementation is a drop-in replacement for the C++ version,
producing the same 'userd' binary installed to /usr/local/bin/userd.

Signed-off-by: Manoharan Sundaramoorthy <manoharan@nexthop.ai>

manoharan-nexthop committed:
(userd) Replace C++ userd with Rust implementation

- Add userd-rs crate with full user management functionality
- Remove C++ userd implementation (userd/ directory)
- Add userd binary to sonic-host-services-rs package
- Add userd.service to sonic-host-services-data package

Features:
- User CRUD operations (create, update, delete)
- Role-based access control (administrator, operator)
- SSH key management with validation
- Security policies and PAM faillock configuration
- Consistency checking on startup
- CONFIG_DB monitoring via SubscriberStateTable

The Rust implementation is a drop-in replacement for the C++ version,
producing the same 'userd' binary installed to /usr/local/bin/userd.

Signed-off-by: Manoharan Sundaramoorthy <manoharan@nexthop.ai>

ronan-nexthop commented:
@manoharan-nexthop can you resolve any open comments that are now resolved, human and AI. @aidan-gallagher could you take a look at this as the 202605 release is getting close and we would like to get this in

mssonicbld commented:
/azp run

manoharan-nexthop commented on userd/src/userd.cpp:
```diff
@@ -0,0 +1,1063 @@
+            } else {
+                SWSS_LOG_DEBUG("User %s already in group %s", username.c_str(), group.c_str());
+            }
+        }
```
yes, this has been handled. (now in Rust code)

manoharan-nexthop commented on userd/src/userd.cpp:
```diff
@@ -0,0 +1,1063 @@
+    bool set_user_password(const std::string& username, const std::string& password_hash) {
+        std::vector<std::string> cmd = {"/usr/sbin/usermod", "-p", password_hash, username};
+
+        if (!SystemCommand::execute(cmd)) {
```
The logging now masks the sensitive information.

mssonicbld commented:
/azp run

aidan-gallagher commented:
> @manoharan-nexthop can you resolve any open comments that are now resolved, human and AI. @aidan-gallagher could you take a look at this as the 202605 release is getting close and we would like to get this in

LGTM
